feat(headless): add insight date facet controller

lbergeron opened the pull request:
https://coveord.atlassian.net/browse/SVCC-2206

This PR builds on top of https://github.com/coveo/ui-kit/pull/2133 and adds the date facet controller for the insight use case.

lbergeron commented on packages/headless/src/controllers/insight/facets/range-facet/date-facet/headless-insight-date-filter.test.ts:
```diff
@@ -0,0 +1,99 @@
+import {buildMockDateFacetRequest} from '../../../../../test/mock-date-facet-request';
```
These tests were copied from `headless-date-filter.test.ts` keeping only tests related to dispatched actions.

aelfaleh commented on packages/headless/src/controllers/insight/facets/range-facet/date-facet/headless-insight-date-facet.ts:
```diff
@@ -0,0 +1,77 @@
+    sortBy(criterion: RangeFacetSortCriterion) {
+      coreController.sortBy(criterion);
+      dispatch(
+        executeSearch(logFacetUpdateSort({facetId: getFacetId(), criterion}))
```
we need to change this event with proper client, in our case `CoveoInsightClient`

aelfaleh replied:
This PR add the new analytics actions https://github.com/coveo/ui-kit/pull/2147

mmitiche commented on packages/headless/src/controllers/insight/facets/range-facet/date-facet/headless-insight-date-facet.ts:
```diff
@@ -0,0 +1,77 @@
+export {buildDateRange};
+
+/**
+ * Creates a `DateFacet` controller instance.
```
```suggestion
 * Creates an `InsightDateFacet` controller instance.
```

mmitiche commented on packages/headless/src/controllers/insight/facets/range-facet/date-facet/headless-insight-date-facet.ts:
```diff
@@ -0,0 +1,77 @@
+ *
+ * @param engine - The headless engine.
+ * @param props - The configurable `DateFacet` controller properties.
+ * @returns A `DateFacet` controller instance.
```
```suggestion
 * @returns An `InsightDateFacet` controller instance.
```

lbergeron replied:
Same here

mmitiche commented on packages/headless/src/controllers/insight/facets/range-facet/date-facet/headless-insight-date-filter.ts:
```diff
@@ -0,0 +1,87 @@
+};
+
+/**
+ * Creates a `DateFilter` controller instance.
```
```suggestion
 * Creates an `InsightDateFilter` controller instance.
```

lbergeron replied:
Same here

mmitiche commented on packages/headless/src/controllers/insight/facets/range-facet/date-facet/headless-insight-date-filter.ts:
```diff
@@ -0,0 +1,87 @@
+ * Creates a `DateFilter` controller instance.
+ * @param engine - The headless engine.
+ * @param props - The configurable `DateFilter` controller properties.
+ * @returns A `DateFilter` controller instance.
```
```suggestion
 * @returns An `InsightDateFilter` controller instance.
```

lbergeron replied:
Same here

mmitiche commented:
Are we planning to export the new insight controllers using the the `insight.index.ts` file only when all the controllers are ready?

nathanlb commented:
> Are we planning to export the new insight controllers using the the `insight.index.ts` file only when all the controllers are ready?

Yep

btaillon commented on packages/headless/src/controllers/insight/facets/range-facet/date-facet/headless-insight-date-facet.ts:
```diff
@@ -0,0 +1,77 @@
+ * @param props - The configurable `DateFacet` controller properties.
+ * @returns A `DateFacet` controller instance.
+ */
+export function buildInsightDateFacet(
```
Same as https://github.com/coveo/ui-kit/pull/2156/files#r908430309 (from another PR)

btaillon commented on packages/headless/src/state/insight-app-state.ts:
```diff
@@ -31,5 +32,6 @@ export type InsightAppState = InsightSearchParametersState &
   FacetSection &
   FacetSearchSection &
   FacetOptionsSection &
+  DateFacetSection &
```
Looks like this might belong in `InsightSearchParametersState`

lbergeron commented:
> TODO: Found an issue when using a timeframe facet with relative timeframes. It apparently caused a malformed search request yielding a 500 error. To investigate.

The issue was that the date facet request and response were not mapped correctly to translate timeframes from relative to absolute format. I modified slightly the `search-mappings` types to be able to use it for insight as well, and updated the insight execute search action accordingly.
